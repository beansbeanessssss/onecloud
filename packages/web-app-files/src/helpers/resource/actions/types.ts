export enum TransferType {
  COPY,
  MOVE,
  DUPLICATE
}
