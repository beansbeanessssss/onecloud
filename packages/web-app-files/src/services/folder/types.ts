export interface FolderLoaderOptions {
  shareIndicators?: boolean
}
