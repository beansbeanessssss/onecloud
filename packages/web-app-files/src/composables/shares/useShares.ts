import { computed, Ref } from 'vue'
import { Share } from '@ownclouders/web-client/src/helpers/share'
import { useStore } from '@ownclouders/web-pkg'

export function useShares() {
  const store = useStore()
  const outgoingLinks: Ref<Share[]> = computed(() => store.getters['Files/outgoingLinks'])
  const outgoingCollaborators: Ref<Share[]> = computed(() =>
    store.getters['Files/outgoingCollaborators'].filter((s) => s.collaborator.displayName)
  )
  return { outgoingCollaborators, outgoingLinks }
}
