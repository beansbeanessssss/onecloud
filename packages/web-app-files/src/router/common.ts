import { RouteComponents } from './router'
import { Location, RouteConfig } from 'vue-router'
import { createLocation, $gettext, isLocationActiveDirector } from './utils'

type commonTypes = 'files-common-favorites' | 'files-common-trash'

export const createLocationCommon = (name: commonTypes, location = {}): Location =>
  createLocation(name, location)

const locationFavorites = createLocationCommon('files-common-favorites')
const locationTrash = createLocationCommon('files-common-trash')

export const isLocationCommonActive = isLocationActiveDirector<commonTypes>(
  locationFavorites,
  locationTrash
)

export const buildRoutes = (components: RouteComponents): RouteConfig[] => [
  {
    path: `/trash`,
    components: {
      app: components.App
    },
    children: [
      {
        name: locationTrash.name,
        path: '',
        component: components.Trashbin,
        meta: {
          hideFilelistActions: true,
          hasBulkActions: true,
          title: $gettext('Deleted files')
        }
      }
    ]
  },
  {
    path: `/favorites`,
    components: {
      app: components.App
    },
    children: [
      {
        name: locationFavorites.name,
        path: '',
        component: components.Favorites,
        meta: {
          hideFilelistActions: true,
          hasBulkActions: false,
          title: $gettext('Favorite files')
        }
      }
    ]
  }
]
