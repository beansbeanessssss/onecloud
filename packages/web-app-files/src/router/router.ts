import Vue, { ComponentOptions } from 'vue'

/**
 * we need to inject the vue files into the route builders,
 * this is because we also import the provided helpers from other js|ts files
 * like mixins, rollup seems to have a problem to import files which contain vue file imports
 * into js files which then again get imported by other vue files...
 */
export interface RouteComponents {
  App: ComponentOptions<typeof Vue>
  Favorites: ComponentOptions<typeof Vue>
  FilesDrop: ComponentOptions<typeof Vue>
  PrivateLink: ComponentOptions<typeof Vue>
  SearchResults: ComponentOptions<typeof Vue>
  PublicLink: ComponentOptions<typeof Vue>
  Shares: {
    SharedWithMe: ComponentOptions<typeof Vue>
    SharedWithOthers: ComponentOptions<typeof Vue>
    SharedViaLink: ComponentOptions<typeof Vue>
  }
  Spaces: {
    DriveResolver: ComponentOptions<typeof Vue>
    Projects: ComponentOptions<typeof Vue>
  }
}
