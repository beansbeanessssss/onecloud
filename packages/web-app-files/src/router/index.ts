import {
  buildRoutes as buildSpacesRoutes,
  isLocationSpacesActive,
  createLocationSpaces
} from './spaces'
import {
  buildRoutes as buildSharesRoutes,
  isLocationSharesActive,
  createLocationShares
} from './shares'
import {
  buildRoutes as buildOperationsRoutes,
  createLocationOperations,
  isLocationOperationsActive
} from './operations'
import {
  buildRoutes as buildCommonRoutes,
  isLocationCommonActive,
  createLocationCommon
} from './common'
import { isRoutePublic, isLocationActive } from './utils'
import { buildRoutes as buildDeprecatedRoutes } from './deprecated'
import { RouteComponents } from './router'
import { RouteConfig } from 'vue-router'

const ROOT_ROUTE = {
  path: '/',
  redirect: (to) => createLocationSpaces(to)
}

const buildRoutes = (components: RouteComponents): RouteConfig[] => [
  ROOT_ROUTE,
  ...buildSharesRoutes(components),
  ...buildSpacesRoutes(components),
  ...buildOperationsRoutes(components),
  ...buildCommonRoutes(components),
  ...buildDeprecatedRoutes()
]

// const isLocationOf = isLocationActiveCreator<locationTypes>(createLocationFor)

export {
  createLocationCommon,
  createLocationShares,
  createLocationSpaces,
  createLocationOperations,
  isLocationOperationsActive,
  isLocationCommonActive,
  isLocationSharesActive,
  isLocationSpacesActive,
  isLocationActive,
  isRoutePublic,
  buildRoutes
}
