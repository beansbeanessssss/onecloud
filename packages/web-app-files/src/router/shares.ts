import { RouteComponents } from './router'
import { Location, RouteConfig } from 'vue-router'
import { createLocation, isLocationActiveDirector, $gettext } from './utils'

type shareTypes =
  | 'files-shares-public-files'
  | 'files-shares-public-link'
  | 'files-shares-public-drop'
  | 'files-shares-private-link'
  | 'files-shares-with-me'
  | 'files-shares-with-others'
  | 'files-shares-via-link'

export const createLocationShares = (name: shareTypes, location = {}): Location =>
  createLocation(name, location)

const locationSharesPublicFiles = createLocationShares('files-shares-public-files')
const locationSharesPublicLink = createLocationShares('files-shares-public-link')
const locationSharesPublicDrop = createLocationShares('files-shares-public-drop')
const locationSharesPrivateLink = createLocationShares('files-shares-private-link')
const locationSharesWithMe = createLocationShares('files-shares-with-me')
const locationSharesWithOthers = createLocationShares('files-shares-with-others')
const locationSharesViaLink = createLocationShares('files-shares-via-link')

export const isLocationSharesActive = isLocationActiveDirector<shareTypes>(
  locationSharesPublicFiles,
  locationSharesPublicLink,
  locationSharesPublicDrop,
  locationSharesPrivateLink,
  locationSharesWithMe,
  locationSharesWithOthers,
  locationSharesViaLink
)

export const buildRoutes = (components: RouteComponents): RouteConfig[] => [
  {
    path: '/shares/public',
    components: {
      app: components.App
    },
    meta: {
      auth: false
    },
    children: [
      {
        name: locationSharesPublicFiles.name,
        path: 'files/:item*',
        component: components.PublicFiles,
        meta: {
          auth: false,
          hasBulkActions: true,
          title: $gettext('Public files'),
          patchCleanPath: true
        }
      }
    ]
  },
  {
    name: locationSharesPublicLink.name,
    path: '/shares/public/link/:token',
    components: {
      fullscreen: components.PublicLink
    },
    meta: {
      auth: false,
      hideHeadbar: true,
      title: $gettext('Resolving public link')
    }
  },
  {
    name: locationSharesPublicDrop.name,
    path: '/shares/public/drop/:token',
    components: {
      app: components.FilesDrop
    },
    meta: { auth: false, title: $gettext('Public file upload') }
  },
  {
    name: locationSharesPrivateLink.name,
    path: '/shares/private/link/:fileId',
    components: {
      fullscreen: components.PrivateLink
    },
    meta: { hideHeadbar: true, title: $gettext('Resolving private link') }
  },
  {
    path: '/shares',
    components: {
      app: components.App
    },
    redirect: locationSharesWithMe,
    children: [
      {
        name: locationSharesWithMe.name,
        path: 'with-me',
        component: components.SharedWithMe,
        meta: {
          hideFilelistActions: true,
          hasBulkActions: true,
          title: $gettext('Files shared with me')
        }
      },
      {
        name: locationSharesWithOthers.name,
        path: 'with-others',
        component: components.SharedWithOthers,
        meta: {
          hideFilelistActions: true,
          hasBulkActions: false,
          title: $gettext('Files shared with others')
        }
      },
      {
        name: locationSharesViaLink.name,
        path: 'via-link',
        component: components.SharedViaLink,
        meta: {
          hideFilelistActions: true,
          hasBulkActions: false,
          title: $gettext('Files shared via link')
        }
      }
    ]
  }
]
