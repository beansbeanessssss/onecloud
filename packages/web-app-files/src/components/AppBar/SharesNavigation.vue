<template>
  <nav id="shares-navigation" class="oc-py-s" :aria-label="$gettext('Shares pages navigation')">
    <oc-list class="oc-flex oc-visible@s">
      <li v-for="navItem in navItems" :key="`shares-navigation-desktop-${navItem.to}`">
        <oc-button
          type="router-link"
          class="oc-mr-s"
          appearance="raw"
          :variation="navItem.active ? 'primary' : 'passive'"
          :to="navItem.to"
        >
          <span v-text="navItem.text" />
        </oc-button>
      </li>
    </oc-list>
    <div class="oc-hidden@s">
      <oc-button id="shares_navigation_mobile" appearance="raw" v-text="$gettext('Shares pages')" />
      <oc-drop toggle="#shares_navigation_mobile" mode="click" close-on-click padding-size="small">
        <oc-list>
          <li v-for="navItem in navItems" :key="`shares-navigation-mobile-${navItem.to}`">
            <oc-button
              type="router-link"
              appearance="raw"
              :variation="navItem.active ? 'primary' : 'passive'"
              :to="navItem.to"
            >
              <span v-text="navItem.text" />
            </oc-button>
          </li>
        </oc-list>
      </oc-drop>
    </div>
  </nav>
</template>

<script>
import {
  isLocationSharesActive,
  locationSharesViaLink,
  locationSharesWithMe,
  locationSharesWithOthers
} from '../../router/shares'
import { computed, getCurrentInstance } from '@vue/composition-api'
import { useRouter } from 'web-pkg/src/composables'

export default {
  setup() {
    const $gettext = getCurrentInstance().proxy.$gettext
    const router = useRouter()
    const sharesRoutes = [
      locationSharesWithMe,
      locationSharesWithOthers,
      locationSharesViaLink
    ].reduce((routes, route) => {
      routes[route.name] = router.getRoutes().find((r) => r.name === route.name)
      return routes
    }, {})
    const navItems = computed(() => [
      {
        to: sharesRoutes[locationSharesWithMe.name].path,
        text: $gettext('Shared with me'),
        active: isLocationSharesActive(router, 'files-shares-with-me')
      },
      {
        to: sharesRoutes[locationSharesWithOthers.name].path,
        text: $gettext('Shared with others'),
        active: isLocationSharesActive(router, 'files-shares-with-others')
      },
      {
        to: sharesRoutes[locationSharesViaLink.name].path,
        text: $gettext('Shared via link'),
        active: isLocationSharesActive(router, 'files-shares-via-link')
      }
    ])
    return {
      navItems
    }
  }
}
</script>
<style lang="scss" scoped>
.router-link-active {
  text-decoration: underline;
}
</style>
