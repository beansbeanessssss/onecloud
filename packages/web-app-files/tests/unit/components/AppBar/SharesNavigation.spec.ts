import SharesNavigation from '../../../../src/components/AppBar/SharesNavigation.vue'
import { locationSharesWithMe } from 'web-app-files/src/router/shares'
import { mock } from 'jest-mock-extended'
import { RouteRecordPublic } from 'vue-router/types/router'
import {
  createStore,
  defaultPlugins,
  defaultStubs,
  shallowMount,
  defaultStoreMockOptions,
  defaultComponentMocks
} from 'web-test-helpers'

const routes = [
  mock<RouteRecordPublic>({
    path: '/files/shares/with-me/',
    name: 'files-shares-with-me'
  }),
  mock<RouteRecordPublic>({
    path: '/files/shares/with-others/',
    name: 'files-shares-with-others'
  }),
  mock<RouteRecordPublic>({
    path: '/files/shares/via-link/',
    name: 'files-shares-via-link'
  })
]

describe('SharesNavigation component', () => {
  it('renders a shares navigation for both mobile and a desktop viewports', () => {
    const { wrapper } = getWrapper()
    expect(wrapper.html()).toMatchSnapshot()
  })
})

function getWrapper({ currentRouteName = locationSharesWithMe.name } = {}) {
  const storeOptions = { ...defaultStoreMockOptions }
  const store = createStore(storeOptions)
  const mocks = defaultComponentMocks({ currentRoute: { name: currentRouteName } })
  mocks.$router.getRoutes.mockImplementation(() => routes)
  return {
    storeOptions,
    mocks,
    wrapper: shallowMount(SharesNavigation as any, {
      global: {
        stubs: defaultStubs,
        renderStubDefaultSlot: true,
        mocks,
        plugins: [...defaultPlugins(), store]
      }
    })
  }
}
