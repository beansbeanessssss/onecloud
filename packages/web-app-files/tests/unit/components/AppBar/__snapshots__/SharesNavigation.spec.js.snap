// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SharesNavigation component renders a shares navigation for both mobile and a desktop viewports 1`] = `
<nav id="shares-navigation" aria-label="Shares pages navigation" class="oc-py-s">
  <ul class="oc-list oc-my-rm oc-mx-rm oc-flex oc-visible@s">
    <li>
      <oc-button-stub type="router-link" size="medium" to="/files/shares/with-me/" submit="button" variation="passive" appearance="raw" justifycontent="center" gapsize="medium" class="oc-mr-s"><span>Shared with me</span></oc-button-stub>
    </li>
    <li>
      <oc-button-stub type="router-link" size="medium" to="/files/shares/with-others/" submit="button" variation="passive" appearance="raw" justifycontent="center" gapsize="medium" class="oc-mr-s"><span>Shared with others</span></oc-button-stub>
    </li>
    <li>
      <oc-button-stub type="router-link" size="medium" to="/files/shares/via-link/" submit="button" variation="passive" appearance="raw" justifycontent="center" gapsize="medium" class="oc-mr-s"><span>Shared via link</span></oc-button-stub>
    </li>
  </ul>
  <div class="oc-hidden@s">
    <oc-button-stub type="button" size="medium" submit="button" variation="passive" appearance="raw" justifycontent="center" gapsize="medium" id="shares_navigation_mobile">Shares pages</oc-button-stub>
    <div id="oc-drop-3" class="oc-drop oc-box-shadow-medium oc-rounded">
      <div class="oc-card oc-card-body oc-rounded oc-background-secondary oc-p-s">
        <ul class="oc-list oc-my-rm oc-mx-rm">
          <li>
            <oc-button-stub type="router-link" size="medium" to="/files/shares/with-me/" submit="button" variation="passive" appearance="raw" justifycontent="center" gapsize="medium"><span>Shared with me</span></oc-button-stub>
          </li>
          <li>
            <oc-button-stub type="router-link" size="medium" to="/files/shares/with-others/" submit="button" variation="passive" appearance="raw" justifycontent="center" gapsize="medium"><span>Shared with others</span></oc-button-stub>
          </li>
          <li>
            <oc-button-stub type="router-link" size="medium" to="/files/shares/via-link/" submit="button" variation="passive" appearance="raw" justifycontent="center" gapsize="medium"><span>Shared via link</span></oc-button-stub>
          </li>
        </ul>
      </div>
    </div>
  </div>
</nav>
`;
